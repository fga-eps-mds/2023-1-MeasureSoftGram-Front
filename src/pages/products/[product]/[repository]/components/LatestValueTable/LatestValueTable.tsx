import React from 'react';
import { format } from 'date-fns';
import { ptBR } from 'date-fns/locale';

import {
  Box,
  TableContainer,
  Table,
  TableCell,
  TableHead,
  TableRow,
  TableBody,
  Paper,
} from '@mui/material';

import { Historical } from '@customTypes/repository';

  
function createData(
  name: string,
  latestValue: number,
  latestData: string,
) {
  return { name, latestValue, latestData};
}

interface Prop {
    title: string;
    latestValue: Historical[];
}

function LatestValueTable({title, latestValue}: Prop) {    
  
    const rows: any[] = [];
    

    latestValue.results.forEach((item) => {
      const {name} = item;
      const {value, created_at: data} = item.latest;

      rows.push(createData(name, value, format(new Date(data), 'dd/MM/yyyy HH:MM', { locale: ptBR })))
    });

    return (
      <Box marginTop="12px" overflow="auto" maxHeight="200px">
        <TableContainer component={Paper}>
        <Table sx={{ minWidth: 650 }} aria-label="simple table">
            <TableHead>
            <TableRow>
                <TableCell>{title}</TableCell>
                <TableCell align="right">Último valor</TableCell>
                <TableCell align="right">Última atualização</TableCell>
            </TableRow>
            </TableHead>
            <TableBody>
            {rows.map((row) => (
                <TableRow
                key={row.name}
                sx={{ '&:last-child td, &:last-child th': { border: 0 } }}
                >
                <TableCell component="th" scope="row">{row.name}</TableCell>
                <TableCell align="right">{row.latestValue}</TableCell>
                <TableCell align="right">{row.latestData}</TableCell>
                </TableRow>
            ))}
            </TableBody>
        </Table>
        </TableContainer>
      </Box>
      );
}

export default LatestValueTable;