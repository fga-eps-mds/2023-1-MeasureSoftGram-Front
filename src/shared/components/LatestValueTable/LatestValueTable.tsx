import React from 'react';
import { format } from 'date-fns';
import { ptBR } from 'date-fns/locale';
import {
  Box,
  TableContainer,
  Table,
  TableCell,
  TableHead,
  TableRow,
  TableBody,
  Paper,
  Alert,
  Fade,
  Skeleton
} from '@mui/material';

import { useRequestLatestValues } from '@hooks/useRequestLatestValues';
import _ from 'lodash';

interface Prop {
  title: string;
  value: 'characteristics' | 'subcharacteristics' | 'measures' | 'metrics';
}

function LatestValueTable({ title, value }: Prop) {
  const { data, error, isLoading, isEmpty } = useRequestLatestValues(value);
  const tableRows: { name: string; latestValue: number; latestData: string }[] = [];

  _.forEach(data, (item) => {
    const { name } = item;
    const { value: latestValue, created_at: latestData } = item.latest;

    tableRows.push({
      name,
      latestValue,
      latestData: format(new Date(latestData), 'dd/MM/yyyy HH:mm:ss', { locale: ptBR })
    });
  });

  return isLoading ? (
    <Skeleton sx={{ width: '100%', minWidth: 400, height: '215px' }} />
  ) : (
    <Box marginTop="12px">
      <TableContainer component={Paper} sx={{ minWidth: 350, maxHeight: '215px' }}>
        <Table stickyHeader aria-label="simple table">
          <TableHead>
            <TableRow>
              <TableCell>{title}</TableCell>
              <TableCell align="right">Último valor</TableCell>
              <TableCell align="right">Última medição</TableCell>
            </TableRow>
          </TableHead>
          <TableBody>
            {tableRows.map((row) => (
              <TableRow key={row.name} sx={{ '&:last-child td, &:last-child th': { border: 0 } }}>
                <TableCell component="th" scope="row">
                  {row.name}
                </TableCell>
                <TableCell align="right">{row.latestValue.toFixed(2)}</TableCell>
                <TableCell align="right" sx={{ width: '250px' }}>
                  {row.latestData}
                </TableCell>
              </TableRow>
            ))}
          </TableBody>
        </Table>
      </TableContainer>
      {error && (
        <Fade in>
          <Alert variant="standard" severity="error">
            Ocorreu um erro ao tentar carregar as informações
          </Alert>
        </Fade>
      )}
      {isEmpty && (
        <Fade in>
          <Alert variant="standard" severity="warning">
            Não há dados para serem exibidos
          </Alert>
        </Fade>
      )}
    </Box>
  );
}

export default LatestValueTable;
