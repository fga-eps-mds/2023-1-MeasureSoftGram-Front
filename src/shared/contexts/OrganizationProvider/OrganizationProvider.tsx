import React, { createContext, useState, useContext, ReactNode, useMemo, useEffect, useCallback } from 'react';

import { Organization } from '@customTypes/organization';
import { organization } from '@services/organization';

interface Props {
  children: ReactNode;
}

interface IOrganizationContext {
  currentOrganization: Organization;
  setCurrentOrganization: (organization: Organization) => void;
  organizationList: Organization[];
  setOrganizationList: (organizations: Organization[]) => void;
}

const OrganizationContext = createContext<IOrganizationContext | undefined>(undefined);

export function OrganizationProvider({ children }: Props) {
  const [organizationList, setOrganizationList] = useState<Organization[]>([]);
  const [currentOrganization, setCurrentOrganization] = useState<Organization>();

  const loadAllOrganization = async () => {
    try {
      const result = await organization?.getAllOrganization();
      setOrganizationList(result?.data?.results);
    } catch (error) {
      console.error(error);
    }
  };

  useEffect(() => {
    loadAllOrganization();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);
  const value = useMemo(
    () => ({
      currentOrganization,
      setCurrentOrganization,
      organizationList,
      setOrganizationList
    }),
    [organizationList, currentOrganization]
  );

  return <OrganizationContext.Provider value={value}>{children}</OrganizationContext.Provider>;
}

export function useOrganizationContext() {
  const context = useContext(OrganizationContext);

  if (context === undefined) {
    throw new Error('OrganizationContext must be used within a OrganizationContext');
  }

  return context;
}
