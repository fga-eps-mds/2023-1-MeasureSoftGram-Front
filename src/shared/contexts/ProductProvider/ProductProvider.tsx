import React, { createContext, useState, useContext, ReactNode, useMemo, useCallback, useEffect } from 'react';

import { Product } from '@customTypes/product';
import { productQuery } from '@services/product';
import { useOrganizationContext } from '@contexts/OrganizationProvider';

interface Props {
  children: ReactNode;
}

interface IProductContext {
  currentProduct?: Product;
  setCurrentProduct: (product: Product) => void;
  productsList?: Product[];
  updateProductList: (products: Product[]) => void;
}

const ProductContext = createContext<IProductContext | undefined>(undefined);

export function ProductProvider({ children }: Props) {
  const [currentProduct, setCurrentProduct] = useState<Product | undefined>();
  const [productsList, setProductsList] = useState<Product[]>([]);

  const { currentOrganization } = useOrganizationContext();

  const updateProductList = useCallback((products: Product[]) => {
    setProductsList(products);
  }, []);

  const loadAllProducts = async () => {
    try {
      const result = await productQuery.getAllProducts(currentOrganization.id);

      updateProductList(result.data.results);
    } catch (error) {
      console.error(error);
    }
  };

  useEffect(() => {
    if (currentOrganization) {
      loadAllProducts();
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [currentOrganization]);

  const value = useMemo(
    () => ({
      currentProduct,
      setCurrentProduct,
      productsList,
      updateProductList
    }),
    [currentProduct, productsList, updateProductList]
  );

  return <ProductContext.Provider value={value}>{children}</ProductContext.Provider>;
}

export function useProductContext() {
  const context = useContext(ProductContext);

  if (context === undefined) {
    throw new Error('useProductContext must be used within a ProductContext');
  }

  return context;
}
