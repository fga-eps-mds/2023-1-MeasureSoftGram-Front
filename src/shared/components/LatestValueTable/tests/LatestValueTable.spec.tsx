import React from 'react';
import { OrganizationProvider } from '@contexts/OrganizationProvider';
import { ProductProvider } from '@contexts/ProductProvider';
import { RepositoryProvider } from '@contexts/RepositoryProvider';
import LatestValueTable from '@components/LatestValueTable';
import { render } from '@testing-library/react';

interface Props {
  children: React.ReactNode;
}

const AllTheProviders = ({ children }: Props) => (
  <OrganizationProvider>
    <ProductProvider>
      <RepositoryProvider>{children}</RepositoryProvider>
    </ProductProvider>
  </OrganizationProvider>
);

describe('<LatestValueTable />', () => {
  test('Should render the LatestValueTable component', () => {
    const { container } = render(<LatestValueTable title="CaracterÃ­ticas" value="characteristics" />, {
      wrapper: AllTheProviders
    });

    expect(container).toMatchSnapshot();
  });
});
