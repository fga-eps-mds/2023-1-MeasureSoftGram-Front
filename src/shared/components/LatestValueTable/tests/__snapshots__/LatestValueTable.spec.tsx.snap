// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LatestValueTable /> Should render the LatestValueTable component 1`] = `
<div>
  <span
    class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1f4rdup-MuiSkeleton-root"
  />
</div>
`;
